### DISCUSSION TITLE
[犀牛鸟 #6790] Qwen3-ASR 的 ncnn 移植

### DISCUSSION THREAD

**Discussion by linkeLi0421** [Tencent/ncnn#6805] [General]
# Qwen3-ASR 转 ncnn：实现与分层验证记录

对应 issue：[Tencent/ncnn#6790](https://github.com/Tencent/ncnn/issues/6790)

- 实现分支：[linkeLi0421/ncnn_llm@qwen3-asr-export](https://github.com/linkeLi0421/ncnn_llm/tree/qwen3-asr-export)
- 实验记录：[linkeLi0421/Qwen3-ASR-ncnn](https://github.com/linkeLi0421/Qwen3-ASR-ncnn)

## 1. 当前结论

`Qwen/Qwen3-ASR-0.6B` 的主要神经网络模块可以通过 pnnx 转为 ncnn，并由 C++
runtime 完成音频前处理、prompt 组装和自回归解码。当前实现不修改 ncnn core。

短中文和较长中文 fixture 的最终 normalized text 已与官方 PyTorch
`Qwen3ASRModel.transcribe()` 对齐；中英数字混合 fixture 中，ncnn 和 PyTorch
都输出 `OpenAPI`，未满足人工 expected `OpenAI API`，因此不能判定为 ncnn
独有的转换错误。

目前应视为可运行的验证原型，还不能宣称完成 #6790：raw tensor 数值误差表、
真人中文音频、Windows 和真实 GPU/Vulkan 路径仍待补齐。

## 2. 实现方式

Qwen3-ASR 的官方 `transcribe()` 不只是神经网络前向，还包含音频读取、重采样、
log-mel、feature mask、prompt/audio token 构造、generation wrapper 和后处理。
因此当前把可静态化的计算拆成四个 ncnn 模块：

- `audio_encoder`
- `text_embed`
- `text_backbone`
- `lm_head`

C++ runtime 负责：

- 读取 16 kHz mono PCM16 WAV 并计算 128-bin log-mel；
- 构造 Qwen3-ASR prompt，将 audio embedding 放入 audio token 位置；
- greedy decode 和 KV cache；
- 输出文本、JSON、mel summary 和模块级调试数据。

## 3. 长音频问题与修复

早期实现把音频拆成固定 256-frame 窗口独立识别，再做 overlap 和文本拼接。实验发现，
即使把同样的短窗口交给 PyTorch 原版逐块识别，也会出现相同的局部错字。这说明问题
主要来自 pipeline 不等价，而不是 ncnn 单个算子的误差。

官方 PyTorch 对 18.78 秒的长中文 fixture 使用整段输入：

```text
input_features: [1, 128, 1878]
audio_features: [244, 1024]
prompt_len: 262
```

旧 `text_seq_len=128` 无法容纳这条路径。因此重新导出
`audio_frames=1878`、`text_seq_len=512`、`kv_cache_len=511`，并在 C++ runtime
中加入 KV cache。另一个关键修复是：pnnx 导出的 prefill K/V cache 带有静态
512 长度，decode 前必须裁到真实 `prompt_len=262`，否则注意力会读取 padding cache。

修复后，长中文 fixture 的 ncnn normalized text 与 PyTorch 对齐。

## 4. 分层验证结果

测试包含三类固定 fixture：

- 短中文普通话；
- 18.78 秒较长中文；
- 中文夹英文、数字和标点。

Linux VM 上使用同一批音频，对比官方 PyTorch baseline 和 ncnn CPU runtime：

| fixture | ncnn strict | PyTorch strict | chunks | RTF | wall time | peak RSS | 结论 |
| --- | --- | --- | ---: | ---: | ---: | ---: | --- |
| `zh_short_tts` | PASS | PASS | 1 | 10.00 | 40.61s | 19.37 GiB | 最终文本对齐 |
| `zh_long_tts` | PASS | PASS | 1 | 3.82 | 72.98s | 19.37 GiB | 长音频整窗和 KV cache 路径对齐 |
| `zh_mixed_tts` | FAIL | FAIL | 1 | 5.66 | 52.39s | 19.38 GiB | 双方均输出 `OpenAPI`，fixture expected 需调整 |

这里 strict normalized 是正式文本契约；宽松的 semantic normalization 只用于分析
空格、标点和英文缩写差异，不替代 strict 结果。

模块级定位也使用了与 ncnn 静态 shape 一致的 PyTorch 路径：
`input_features=[1,128,1878]`、全 1 feature mask、244 个 audio placeholder、
`prompt_len=262`。三条 fixture 的 shape 和首个 greedy token 均已对齐，首 token
分别为 `100644`、`43288`、`100644`。

当前实现已经支持双方导出 `.f32` raw tensor，评估器可计算 `max_abs`、`mean_abs`、
`p99_abs`、cosine similarity 和 selected logits top-5 agreement；三条 fixture
的真实数值表尚需在 VM 上重跑。

## 5. 平台与性能说明

- macOS arm64（Apple M1 Pro）：CPU-only CMake build/smoke 已通过。
- Linux VM：CMake build 和 ncnn CPU 端到端验证已通过。
- VM 中 RTX 4090 可由 `nvidia-smi` 看到，但 Vulkan 当前只枚举到 `llvmpipe`，
  所以上表是 CPU 结果，不代表 4090/Vulkan 性能。
- 原始 ncnn build 在该 VM 上触发 `get_data_cache_size()` 的 CPU cache parsing
  `SIGFPE`；本次 Linux 验证使用临时 patched build 绕过该问题。

long-window fp32 静态模型的内存代价较高，当前 CPU 峰值 RSS 约 19.4 GiB。KV cache
解决了逐 token 重跑完整 text backbone 的问题，但固定大 shape 的模型体积和内存仍需优化。

## 6. 最小复测入口

```bash
qwen3_asr_main \
  --model <qwen3-asr-ncnn-runtime-dir> \
  --audio-wav sample.wav \
  --generate-from-features \
  --text-out out.txt \
  --json-out out.json \
  --dump-mel-summary mel.json
```

实现分支中的 `tests/qwen3_asr` 还提供三类 fixture、PyTorch baseline、模块 raw dump、
normalized text 对比和平台 smoke 的脚本。
